src/Datastore/Operation.php

```diff
@@ -665,4 +665,30 @@ private function readOptions(array $options = [])
             'readOptions' => $readOptions
         ]);
     }
+
+    /**
+     * Sort entities into the order given in $keys.
+     *
+     * @param Entity[] $entities
+     * @param Key[] $keys
+     * @return Entity[]
+     */
+    private function sortEntities(array $entities, array $keys)
+    {
+        $res = [];
+        foreach ($keys as $key) {
+            $path = $key->path();
+
+            foreach ($entities as $arrayKey => $entity) {
+                $entityPath = $entity->key()->path();
+
+                if ($path === $entityPath) {
+                    $res[] = $entity;
+                    unset($entities[$arrayKey]);
```

Comment:
Sorry for chiming in late. This code will slow when there are many entities. Probably it's ok because the limit of number of batch lookup is 1000. Have you considered to make it linear?


```php

$map = [];
foreach ($entities as $entity) {
  map[(string) $entity->key()] = $entity;
}
$ret = [];
foreach ($keys as $key) {
  $ret[] = map[(string) $key];
}
return $ret;
```

WDYT?